nexus/pack.cc

```diff
@@ -2304,6 +2304,128 @@ struct NexusPacker
+        size_t num_ff   = 0;
+        size_t num_pair = 0;
+        size_t num_glue = 0;
+
+        for (auto &cell : ctx->cells) {
+            CellInfo *ff = cell.second.get();
+            if (ff->type != id_OXIDE_FF) {
+                continue;
+            }
+
+            num_ff++;
+
+            // Get input net
+            // At the packing stage all inputs go to M
+            NetInfo *di = get_net_or_empty(ff, id_M);
+            if (di == nullptr || di->driver.cell == nullptr) {
+                continue;
+            }
+
+            // Skip if there are multiple sinks on that net
+            if (di->users.size() != 1) {
+                continue;
+            }
+
+            // Check if the driver is a LUT and the direct connection is from F
+            CellInfo* lut = di->driver.cell;
+            if (lut->type != id_OXIDE_COMB) {
+                continue;
+            }
+            if (di->driver.port != id_F  &&
+                di->driver.port != id_OFX)
+            {
+                continue;
+            }
+
+            // The FF must not use M and DI at the same time
+            if (get_net_or_empty(ff, id_DI)) {
+                continue;
+            }
+
+            // The LUT must be in LOGIC/CARRY mode
+            if (str_or_default(lut->params, id_MODE, "LOGIC") != "LOGIC" &&
+                str_or_default(lut->params, id_MODE, "LOGIC") != "CCU2") {
+                continue;
+            }
+
+            // The FF cannot be in another cluster
+            if (ff->cluster != ClusterId()) {
+                continue;
+            }
+
+            // A free LUT, create a new cluster
+            if (lut->cluster == ClusterId()) {
+
+                lut->cluster = lut->name;
+                lut->constr_children.push_back(ff);
+
+                ff->cluster = lut->name;
+                ff->constr_x = 0;
+                ff->constr_y = 0;
+                ff->constr_z = 2;
+                ff->constr_abs_z = false;
+
+                num_pair++;
+            }
+            // Attach the FF to the existing cluster of the LUT
+            else {
+
+                // No order not to make too large carry clusters pack only the
+                // given fraction of FFs there.
+                if(str_or_default(lut->params, id_MODE, "LOGIC") == "CCU2") {
+                    float r = (float)(ctx->rng() % 1000) * 1e-3f;
+                    if (r > carry_ratio) {
+                        continue;
+                    }
+                }
+
+                // Get the cluster root
+                CellInfo* root = ctx->cells.at(lut->cluster).get();
+
+                // Constrain the FF relative to the LUT
+                ff->cluster = root->cluster;
+                ff->constr_x = lut->constr_x;
+                ff->constr_y = lut->constr_y;
+                ff->constr_z = lut->constr_z + 2;
+                ff->constr_abs_z = lut->constr_abs_z;
+                root->constr_children.push_back(ff);
```

Comment:
this case also needs to be checking that the FF can legally be added to the cluster (i.e. CLK/LSR/CE sharing etc)

Reply:
Oh, you are right. I missed that... I'll push a fix soon.

Reply:
Alright, should be fixed by now.